optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/AbstractSolution.java

```diff
@@ -16,53 +16,52 @@
 
 package org.optaplanner.core.impl.domain.solution;
 
+import org.optaplanner.core.api.domain.solution.*;
+import org.optaplanner.core.api.score.Score;
+import org.optaplanner.core.impl.domain.common.ReflectionHelper;
+
 import java.io.Serializable;
 import java.lang.reflect.Field;
 import java.lang.reflect.Method;
 import java.util.ArrayList;
 import java.util.Collection;
-import java.util.List;
 import java.util.Map;
 
-import org.optaplanner.core.api.domain.solution.PlanningEntityCollectionProperty;
-import org.optaplanner.core.api.domain.solution.PlanningEntityProperty;
-import org.optaplanner.core.api.domain.solution.Solution;
-import org.optaplanner.core.api.domain.solution.cloner.DeepPlanningClone;
-import org.optaplanner.core.api.score.Score;
-import org.optaplanner.core.impl.domain.common.ReflectionHelper;
-
 /**
  * Currently only used by OptaPlanner Workbench.
  * TODO Should we promote using this class in the examples and docs too?
  * We can never enforce it, as the user might want to use a different superclass.
  * @param <S> the {@link Score} type used by this use case
  */
-public abstract class AbstractSolution<S extends Score> implements Solution<S>, Serializable {
+public abstract class AbstractSolution<S extends Score> implements Serializable {
 
     protected S score;
 
-    @Override
+    @PlanningScore
     public S getScore() {
         return score;
     }
 
-    @Override
     public void setScore(S score) {
         this.score = score;
     }
 
-    @Override
-    public Collection<?> getProblemFacts() {
-        List<Object> factList = new ArrayList<Object>();
+    /**
+     * Convenience method for tests.
+     *
+     * @return All entities from anywhere in this class' hierarchy.
+     */
+    protected Collection<?> getProblemFacts() {
         Class<? extends AbstractSolution> instanceClass = getClass();
-        addProblemFactsFromClass(factList, instanceClass);
-        return factList;
+        return getProblemFactsFromClass(instanceClass);
     }
 
-    private void addProblemFactsFromClass(List<Object> factList, Class<?> instanceClass) {
+    @PlanningFactCollectionProperty
+    private Collection<Object> getProblemFactsFromClass(Class<?> instanceClass) {
```

Comment:
Doesn't the @PlanningFactCollectionProperty need to be on getProblemFacts()?

If the  @PlanningFactCollectionProperty is on a method that contains 1 or more arguments, OptaPlanner must fail-fast during buildSolver() with a clear error message. We have a test for AbstractSolution, so I presume that fail-fast might be missing. 

Some btw if @PlanningFactCollectionProperty is on a method that doesn't return a collection: that should also fail fast.